### ISSUE TITLE
MeanVarianceNormalization without division by zero

### ISSUE THREAD

**Issue by enpasos** [onnx/onnx#3947]
# Bug Report

### Is the issue related to model conversion?
Yes: While converting LayerNorm to onnx using MeanVarianceNormalization

### Describe the bug
Without Variance in the input data I get out NaN. But the outcome should be zero. Look like division by zero. 

Maybe it works as designed, then this is more a feature request for a parameter epsilon: epsilon = 0 as now, and with some small epsilon division by epsilon instead of zero..

### System information
- OS Platform and Distribution (Windows 11): 
- ONNX version:  1.10.2
- Protobuf version: 3

**Comment by WilliamTambellini** [comment]
Well apparently the original Layer Norm did nt propose to add an epsilon:
https://arxiv.org/pdf/1607.06450.pdf
Now in "real life", most add :
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html
There is a batchNorm op
https://github.com/onnx/onnx/blob/main/docs/Operators.md#BatchNormalization
a meanVarNorm op:
https://github.com/onnx/onnx/blob/main/docs/Operators.md#MeanVarianceNormalization
but no layerNorm ?

**Comment by gramalingam** [comment]
For MeanVarianceNormalization, adding an epsilon attribute makes sense.

**Comment by fdwr** [comment]
> For MeanVarianceNormalization, adding an epsilon attribute makes sense.

@gramalingam: Concur.

We encountered the missing epsilon while implementing the [Chromium WebNN backend atop ORT](https://chromium-review.googlesource.com/c/chromium/src/+/6758777). ONNX's other normalization operators have an epsilon parameter. So it's aberrant that just MVN is missing it (which would let us shave off ~100 lines of decomposition code).

- ✅ [`BatchNormalization`](https://onnx.ai/onnx/operators/onnx__BatchNormalization.html)
- ✅ [`GroupNormalization`](https://onnx.ai/onnx/operators/onnx__GroupNormalization.html)
- ✅ [`InstanceNormalization`](https://onnx.ai/onnx/operators/onnx__InstanceNormalization.html)
- ✅ [`LayerNormalization`](https://onnx.ai/onnx/operators/onnx__LayerNormalization.html)
- ❌ [`MeanVarianceNormalization`](https://onnx.ai/onnx/operators/onnx__MeanVarianceNormalization.html)
- ✅ [`RMSNormalization`](https://onnx.ai/onnx/operators/onnx__RMSNormalization.html)

Reopening (rather than creating a new one).

**Comment by ogulcanaydogan** [comment]
Hi! I'd like to work on this. Based on the discussion, the approach would be:

1. Add an `epsilon` attribute to `MeanVarianceNormalization` in a new opset version
2. Default `epsilon = 0` for backward compatibility
3. When `epsilon > 0`, use it as the divisor floor to prevent division by zero
4. Update the reference implementation and add test cases for the zero-variance edge case

This aligns with how PyTorch's `LayerNorm` and other frameworks handle it. I'll also check if the ONNX spec has a standard pattern for introducing new optional attributes to existing ops.

Does this approach look right?

**Comment by enpasos** [comment]
Thanks for picking this up — this seems like a sensible direction to me.

It may help to make a few details explicit in the proposal: preserving current MVN behavior by default in the new opset for backward compatibility, stating the exact epsilon formula in the spec, and adding tests for zero-variance and near-zero-variance inputs.

With that clarified, this looks good to me.

**Comment by ogulcanaydogan** [comment]
Thanks for the feedback @enpasos! Here's the explicit proposal:

## Proposal: Add `epsilon` attribute to MeanVarianceNormalization

### Current state
MVN's function body already hardcodes `Epsilon = Constant <value = float {1e-9}>()` (see `onnx/defs/nn/defs.cc:2204`), but this value is not user-configurable. Meanwhile, every other normalization op (BatchNorm, LayerNorm, InstanceNorm, GroupNorm, RMSNorm) exposes `epsilon` as an attribute.

### Changes

**New attribute:**
```
epsilon : float (default = 1e-9)
```

Using `1e-9` as default preserves exact numerical parity with the current function body. (Other normalization ops default to `1e-5`, but changing the default would alter existing model outputs.)

**Epsilon placement — keep current pattern (add to std, after sqrt):**
```
Current:  Y = (X - mean) / (sqrt(variance) + 1e-9)      ← hardcoded
Proposed: Y = (X - mean) / (sqrt(variance) + epsilon)    ← configurable
```

Note: Other normalization ops add epsilon *before* sqrt: `sqrt(variance + epsilon)`. However, changing the formula structure would break backward compatibility. If the community prefers aligning with other ops, we could do that with `default = 0` instead (since `sqrt(var + 0) ≈ sqrt(var) + 0`), but this changes the formula semantics.

**Spec text update:**
```
(X-EX) / (sqrt(E(X-EX)^2) + epsilon)
```

**Function body change** (in `defs.cc`):
```diff
- Epsilon = Constant <value = float {1e-9}>()
+ Epsilon = Constant <value_float: float = @epsilon>()
```

**Opset version:** Bump from 13 to the next available opset.

**Files to change:**
- `onnx/defs/nn/defs.cc` — Add epsilon attribute, update function body
- `onnx/defs/nn/old.cc` — Move current ver 13 definition here
- `onnx/defs/operator_sets.h` — Register new version
- `onnx/version_converter/convert.h` — Add compatible adapter (13→new)
- `onnx/backend/test/case/node/meanvariancenormalization.py` — Add test cases for:
  - Default epsilon (backward compat)
  - Custom epsilon value
  - Zero-variance input (division by zero edge case)
  - Near-zero-variance input
- Auto-generated docs (Operators.md, Changelog.md) via `python onnx/defs/gen_doc.py`

**Backward compatibility:** Models using MVN with default `epsilon=1e-9` produce identical results to current behavior. The only change is that users can now explicitly set a different epsilon value.

@enpasos @fdwr — does this look right? Happy to proceed with implementation once confirmed.

**Comment by enpasos** [comment]
Thanks for clarifying — this looks reasonable to me.

Keeping the current MVN formula structure and using `epsilon = 1e-9` as the default seems sensible for backward compatibility.

**Comment by fdwr** [comment]
@ogulcanaydogan Sounds good 👍. 1e-9 is a fine default. TY.

@huningxin: FYI, adding the `epsilon` attribute may let us delete dozens of lines of emulation from https://chromium-review.googlesource.com/c/chromium/src/+/6758777.

**Comment by CuriousPan** [comment]
I can pick and finalize the changes.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

From the report, the useful first step is to pin down the exact code path involved before touching anything — these tend to have a small true cause behind a broad-looking symptom, and guessing at it usually produces a patch that masks the behavior instead of fixing it.

Plan: trace the path, reproduce it, then make a focused change with a test if this area has coverage. I'll keep the diff small so it's easy to review.

If there's an existing convention in this repo for how this kind of thing is handled, I'd rather follow it than invent my own — happy to be pointed at one.

(claiming via @LeonxLJX)
